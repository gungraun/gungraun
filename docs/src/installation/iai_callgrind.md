# Gungraun

Gungraun is divided into the library `iai-callgrind` and the benchmark runner
`iai-callgrind-runner`.

## Installation of the library

To start with Gungraun, add the following to your `Cargo.toml` file:

```toml
[dev-dependencies]
iai-callgrind = "0.16.1"
```

or run

```bash
cargo add --dev iai-callgrind@0.16.1
```

## Installation of the benchmark runner

To be able to run the benchmarks you'll also need the `iai-callgrind-runner`
binary installed somewhere in your `$PATH`. Otherwise, there is no need to
interact with `iai-callgrind-runner` as it is just an implementation detail.

### From Source

```shell
cargo install --version 0.16.1 iai-callgrind-runner
```

There's also the possibility to install the binary somewhere else and point the
`IAI_CALLGRIND_RUNNER` environment variable to the absolute path of the
`iai-callgrind-runner` binary like so:

```shell
cargo install --version 0.16.1 --root /tmp iai-callgrind-runner
IAI_CALLGRIND_RUNNER=/tmp/bin/iai-callgrind-runner cargo bench --bench my-bench
```

### Binstall

The `iai-callgrind-runner` binary is
[pre-built](https://github.com/gungraun/gungraun/releases/tag/v0.16.1)
for most platforms supported by valgrind and easily installable with
[binstall](https://github.com/cargo-bins/cargo-binstall)

```shell
cargo binstall iai-callgrind-runner@0.16.1
```

## Updating

When updating the `iai-callgrind` library, you'll also need to update
`iai-callgrind-runner` and vice-versa or else the benchmark runner will exit
with an error.

### In the Github CI

Since the `iai-callgrind-runner` version must match the `iai-callgrind` library
version it's best to automate this step in the CI. A job step in the github
actions CI could look like this

```yaml
- name: Install iai-callgrind-runner
  run: |
    version=$(cargo metadata --format-version=1 |\
      jq '.packages[] | select(.name == "iai-callgrind").version' |\
      tr -d '"'
    )
    cargo install iai-callgrind-runner --version $version
```

Or, speed up the overall installation time with `binstall` using the
[taiki-e/install-action](https://github.com/taiki-e/install-action)

```yaml
- uses: taiki-e/install-action@cargo-binstall
- name: Install iai-callgrind-runner
  run: |
    version=$(cargo metadata --format-version=1 |\
      jq '.packages[] | select(.name == "iai-callgrind").version' |\
      tr -d '"'
    )
    cargo binstall --no-confirm iai-callgrind-runner --version $version
```
