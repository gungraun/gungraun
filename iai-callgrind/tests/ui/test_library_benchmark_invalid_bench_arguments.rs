use iai_callgrind::library_benchmark;

// missing argument of the benchmark
#[library_benchmark]
#[bench::id(42)]
#[benches::multi(42)]
fn bench10() {}

// missing argument of the bench attribute
#[library_benchmark]
#[bench::id()]
fn bench20(my: i32) {}

#[library_benchmark]
#[benches::multi(args = [])]
fn bench25(my: i32) {}

// too many arguments of the bench attribute
#[library_benchmark]
#[bench::id(42, 8)]
fn bench30(my: i32) {}

#[library_benchmark]
#[benches::multi((42, 8))]
fn bench30(my: i32) {}

// incorrect argument type
#[library_benchmark]
#[bench::id("hello")]
fn bench40(my: u8) {}

#[library_benchmark]
#[benches::multi("hello")]
fn bench45(my: u8) {}

// incorrect return type
#[library_benchmark]
#[bench::id(42)]
fn bench5(my: u8) -> String {
    my
}

fn main() {}
