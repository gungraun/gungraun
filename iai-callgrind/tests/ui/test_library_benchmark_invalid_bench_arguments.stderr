error: Expected 0 arguments but found 1

         = help: This argument is expected to have the same amount of parameters as the benchmark function

 --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:5:13
  |
5 | #[bench::id(42)]
  |             ^^

error: Expected 0 arguments but found 1

         = help: This argument is expected to have the same amount of parameters as the benchmark function

 --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:6:18
  |
6 | #[benches::multi(42)]
  |                  ^^

error: Expected 1 arguments but found 0

         = help: This argument is expected to have the same amount of parameters as the benchmark function

  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:10:1
   |
10 | #[library_benchmark]
   | ^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `library_benchmark` (in Nightly builds, run with -Z macro-backtrace for more info)

error: Expected 1 arguments but found 0

         = help: This argument is expected to have the same amount of parameters as the benchmark function

  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:14:1
   |
14 | #[library_benchmark]
   | ^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `library_benchmark` (in Nightly builds, run with -Z macro-backtrace for more info)

error: Expected 1 arguments but found 2

         = help: This argument is expected to have the same amount of parameters as the benchmark function

  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:20:13
   |
20 | #[bench::id(42, 8)]
   |             ^^

error: Expected 1 arguments but found 2

         = help: This argument is expected to have the same amount of parameters as the benchmark function

  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:24:18
   |
24 | #[benches::multi((42, 8))]
   |                  ^^^^^^^

error[E0423]: expected function, found module `bench5::bench5`
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:52:5
   |
52 |     bench5::bench5();
   |     ^^^^^^^^^^^^^^ not a function
   |
note: function `crate::bench5::__iai_callgrind_wrapper_mod::bench5` exists but is inaccessible
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:37:1
   |
37 | #[library_benchmark]
   | ^^^^^^^^^^^^^^^^^^^^ not accessible
   = note: this error originates in the attribute macro `library_benchmark` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0423]: expected function, found module `bench6::bench6`
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:54:5
   |
54 |     bench6::bench6();
   |     ^^^^^^^^^^^^^^ not a function
   |
note: function `crate::bench6::__iai_callgrind_wrapper_mod::bench6` exists but is inaccessible
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:43:1
   |
43 | #[library_benchmark]
   | ^^^^^^^^^^^^^^^^^^^^ not accessible
   = note: this error originates in the attribute macro `library_benchmark` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0423]: expected function, found module `bench7::bench7`
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:56:5
   |
56 |     bench7::bench7();
   |     ^^^^^^^^^^^^^^ not a function
   |
note: function `crate::bench7::__iai_callgrind_wrapper_mod::bench7` exists but is inaccessible
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:46:1
   |
46 | #[library_benchmark]
   | ^^^^^^^^^^^^^^^^^^^^ not accessible
   = note: this error originates in the attribute macro `library_benchmark` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0603]: module import `bench5` is private
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:52:13
   |
52 |     bench5::bench5();
   |             ^^^^^^ private module import
   |
note: the module import `bench5` is defined here...
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:37:1
   |
37 | #[library_benchmark]
   | ^^^^^^^^^^^^^^^^^^^^
note: ...and refers to the module `bench5` which is defined here
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:37:1
   |
37 | #[library_benchmark]
   | ^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the attribute macro `library_benchmark` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0603]: module import `bench6` is private
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:54:13
   |
54 |     bench6::bench6();
   |             ^^^^^^ private module import
   |
note: the module import `bench6` is defined here...
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:43:1
   |
43 | #[library_benchmark]
   | ^^^^^^^^^^^^^^^^^^^^
note: ...and refers to the module `bench6` which is defined here
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:43:1
   |
43 | #[library_benchmark]
   | ^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the attribute macro `library_benchmark` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0603]: module import `bench7` is private
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:56:13
   |
56 |     bench7::bench7();
   |             ^^^^^^ private module import
   |
note: the module import `bench7` is defined here...
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:46:1
   |
46 | #[library_benchmark]
   | ^^^^^^^^^^^^^^^^^^^^
note: ...and refers to the module `bench7` which is defined here
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:46:1
   |
46 | #[library_benchmark]
   | ^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the attribute macro `library_benchmark` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:29:13
   |
29 | #[bench::id("hello")]
   |             ^^^^^^^
   |             |
   |             expected `u8`, found `&str`
   |             arguments to this function are incorrect
   |
help: the return type of this call is `&'static str` due to the type of the argument passed
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:29:13
   |
29 | #[bench::id("hello")]
   |             ^^^^^^^ this argument influences the return type of `std`
note: function defined here
  --> $RUST/core/src/hint.rs
   |
   | pub const fn black_box<T>(dummy: T) -> T {
   |              ^^^^^^^^^

error[E0308]: mismatched types
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:33:18
   |
33 | #[benches::multi("hello")]
   |                  ^^^^^^^
   |                  |
   |                  expected `u8`, found `&str`
   |                  arguments to this function are incorrect
   |
help: the return type of this call is `&'static str` due to the type of the argument passed
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:33:18
   |
33 | #[benches::multi("hello")]
   |                  ^^^^^^^ this argument influences the return type of `std`
note: function defined here
  --> $RUST/core/src/hint.rs
   |
   | pub const fn black_box<T>(dummy: T) -> T {
   |              ^^^^^^^^^

error[E0308]: mismatched types
  --> tests/ui/test_library_benchmark_invalid_bench_arguments.rs:40:5
   |
39 | fn bench5(my: u8) -> String {
   |                      ------ expected `String` because of return type
40 |     my
   |     ^^- help: try using a conversion method: `.to_string()`
   |     |
   |     expected `String`, found `u8`
